File: code/cpp/eagerDjikstra.cpp
#include <iostream>

#include <vector>
#include <string>
#include <algorithm>
#include <limits>
#include "helpers.h"


std::pair<std::vector<int>, std::vector<int>> djikstra(networks::Graph &graph, int start)
{
    std::vector<bool> visited(graph.numNodes, false);
    std::vector<int> distance(graph.numNodes, INT32_MAX);
    std::vector<int> parents(graph.numNodes, -1);
    
    distance[start] = 0;

    networks::IndexedMinPQ ipq = networks::IndexedMinPQ(graph.numNodes);  
    ipq.push(start, 0);

    while (!ipq.isEmpty())
    {
        std::pair<int, int> elem = ipq.pop(); 
        
        int minValue = elem.second, index = elem.first;
        ipq.deleteKey(index);
        visited[index] = true;

        if (distance[index] < minValue) continue;
        for (auto edge : graph.adjList[index])
        {
            if (visited[edge.first]) continue;
            int newDist = distance[index] + edge.second;

            if (newDist < distance[edge.first])
            {
                parents[edge.first] = index;
                distance[edge.first] = newDist;

                if (!ipq.contains(edge.first))
                    ipq.push(edge.first, newDist);
                else
                    ipq.decreaseKey(edge.first, newDist);
            }
        }
    }

    return {distance, parents};
}

std::pair<int, std::vector<int>> shortest_path(networks::Graph &graph, int start, int target)
{
    std::pair<std::vector<int>, std::vector<int>> ans = djikstra(graph, start);
    std::vector<int> path;
    if (ans.first[target] == INT32_MAX) return {-1, path};

    int node = target;
    while (node != -1)
    {
        path.push_back(node);
        node = ans.second[node];
    }
    std::reverse(path.begin(), path.end());
    return {ans.first[target], path};
}


int main()
{
#ifndef FILE_INPUT
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif

    int numNodes, numEdges;
    std::cin >> numNodes >> numEdges;

    bool directed, weighted;
    std::cin >> weighted >> directed;

    int u, v, w;
    std::vector<networks::Edge> edges;
    if (weighted)
    {
        for (int i=0; i<numEdges; i++)
        {
            std::cin >> u >> v >> w;
            edges.push_back(networks::Edge(u, v, w));
        }
    }
    else
    {
        for (int i=0; i<numEdges; i++)
        {
            std::cin >> u >> v;
            edges.push_back(networks::Edge(u, v));
        }
    }

    // Initializing Graph
    networks::Graph graph = networks::Graph(numNodes, edges, directed);

    std::pair<std::vector<int>, std::vector<int>> ans = djikstra(graph, 0);
    helpers::printVector(ans.first, "Distance Array: ");
    helpers::printVector(ans.second, "Parents Array: ");

    std::pair<int, std::vector<int>> ans_sp = shortest_path(graph, 0, 5);
    std::cout << "Shortest Distance: " << ans_sp.first << "\n";
    helpers::printVector(ans_sp.second, "Shortest Path: ");
}

File: code/python/eager_djikstra.py
import helpers as H

# global variables
INPUT_PATH = 'input.txt'
OUTPUT_PATH = 'output.txt'

# clearing the contents of the file if any
open(OUTPUT_PATH, 'w').close()


def djikstra(graph: H.Graph, start: int):
    visited = [False for _ in range(graph.num_nodes)]
    distance = [float('inf') for _ in range(graph.num_nodes)]
    parents = [None for _ in range(graph.num_nodes)]
    
    distance[start] = 0
    ipq = H.IndexedMinPQ(graph.num_nodes)
    ipq.push(start, 0)
    
    while not ipq.is_empty():
        index, min_value = ipq.pop()
        visited[index] = True
        
        if distance[index] < min_value: continue
        
        for edge in graph.adj_list[index]:
            if visited[edge[0]]: continue
        
            new_dist = distance[index] + edge[1]
            if new_dist < distance[edge[0]]:
                parents[edge[0]] = index
                distance[edge[0]] = new_dist
        
                if edge[0] in ipq.heap:
                    ipq.update(edge[0], new_dist)
                else:
                    ipq.push(edge[0], new_dist)
                    
    return distance, parents


def shortest_path(graph: H.Graph, source: int, target: int):
    dist, parents = djikstra(graph, source)
    path = []
    if dist[target] == float('inf'): return path
    
    node = target
    while node is not None:
        path.append(node)
        node = parents[node]

    path = path[::-1]
    return path, dist[target]
        
if __name__ == '__main__':
    with open(INPUT_PATH, 'r') as f:
        num_nodes, num_edges = (int(i) for i in f.readline().split(" "))
        directed, weighted = (int(i) for i in f.readline().split(" "))
        
        edges = []
        if weighted:
            for i in range(num_edges):
                u, v, w = (int(i) for i in f.readline().split(" "))
                edges.append(H.Edge(u, v, w))
        else:
            for i in range(num_edges):
                u, v = (int(i) for i in f.readline().split(" "))
                edges.append(H.Edge(u, v))
                
    graph = H.Graph(num_nodes, edges, directed)                
    
    ans = djikstra(graph, 0)
    ans_sp = shortest_path(graph, 0, 5)
    with open(OUTPUT_PATH, 'a') as f:
        f.write(f"Distance Array: {ans[0]}\nParents Array: {ans[1]}\
                \nShortest Path: {ans_sp[0]}\nCost of Shortest Path: {ans_sp[1]}")
